### ISSUE TITLE
There is not model_best.pth after training finish

### ISSUE THREAD

**Issue by HyuanTan** [Eromera/erfnet_pytorch#11]
Hi, Eromera. There is `--epochs-save` to save model every X epochs, but when I set it to none zero, It can't save the model as I expected. And another issue is that when the training is finished, some time it will save the model_best.pth and someting just get the model_encoder_best.pth.
Thanks!

**Comment by Eromera** [comment]
Hi!
I didn't get what do you mean when you say that you set --epochs-save to none zero, could you say it in other way?
About the other "issue", it is not really an issue: you have two options to train the model:
1) from scratch: train X epochs the encoder alone (saved as model_encoder_best.pth after the X epochs), then the code grabs automatically those encoder weights and creates the full network and starts training the full ERFNet (enc+dec) for another X epochs.
2) pretrained: Loads weights that were trained in ImageNet in order to load a "pretrained encoder". In this case, it only trains the full network (X epochs training full ERFNet) because the encoder filters are already good.
I hope this helps :)

**Comment by HyuanTan** [comment]
Hi, Thanks fro respond.
In the main.py for training, I set the --epochs-save to 50, it seems that the training process didn't save the model every 50 echop.
I know that there are  two options to train the model, and I use "from scratch" to train the model.Sometime, after the ENC and Dec finish, the model_best.pth didn't saved.

Thanks!

**Comment by HyuanTan** [comment]
After training finished using "from scratch", the output just like this and didn't save the best model:

![erfnet-1](https://user-images.githubusercontent.com/14145396/34720135-9d10466e-f578-11e7-9391-c8bdb913f469.png)

![erfnet-2](https://user-images.githubusercontent.com/14145396/34720222-dd575cda-f578-11e7-95fb-0610357e036a.png)

And sometime, it will save the best model like this:
![3](https://user-images.githubusercontent.com/14145396/34721479-dbbe2200-f57d-11e7-99c2-027718c73c2c.png)

![4](https://user-images.githubusercontent.com/14145396/34721501-ede6e5fc-f57d-11e7-95a6-5730c8e28681.png)

It seems sometime "save_checkpoint" don't work?
![save](https://user-images.githubusercontent.com/14145396/34721533-0c2eb79c-f57e-11e7-82f1-7c8abc62b6a2.png)

**Comment by HyuanTan** [comment]
I have found that if I do the "iouVal", only when the IoU in decode  is better than the IoU in encode, the model_best.pth will be saved:

![1-1](https://user-images.githubusercontent.com/14145396/34773165-b18b8112-f645-11e7-9d57-6b6d6870b832.png)

Maybe that is why sometime I can get the model_best.pth and sometime not in my case.
Thanks!!

**Comment by zd19901107** [comment]
You achieved a mean IoU of 77.79% on the val dataset, how do you do that?  Is it the same task with the paper? I run the model, and the result is similar to the result on the paper!

**Comment by HyuanTan** [comment]
@zd19901107  When I train on CityScape dataset, the val IoU is  similar to the paper. The IoU of 77.79% is the result training on my own dataset which is much more simple than CityScape。

**Comment by Eromera** [comment]
@HyuanTan Sorry for taking so long to answer, did you manage to solve the issue with saving model_best.pth?

**Comment by HyuanTan** [comment]
@Eromera Hi，I solved this issue by change some code in main.py.

The previous code:
![image](https://user-images.githubusercontent.com/14145396/36967194-07739ef6-209a-11e8-968a-437fc172de29.png)
When I do the "iouVal" and train from the scratch, only when the IoU in decode is better than the IoU in encode, the model_best.pth will be saved.

But if I change the code like this:
![image](https://user-images.githubusercontent.com/14145396/36967145-e729f2c6-2099-11e8-8a30-73d8a0e96a61.png)
It will save the best model in encode without considering decode.

Moreover, if the "iouVal = False" and train from the scratch, another place need some change:
![image](https://user-images.githubusercontent.com/14145396/36967553-0b4d535e-209b-11e8-8c43-4c9a24379eaf.png)
Combine `is_best = current_acc > best_acc` and current_acc` = -average_epoch_loss_val` we need to save the model which loss is smaller not larger.

If the code I changed in main.py is wrong, please tell me. Thanks!!

**Comment by Eromera** [comment]
Hey @HyuanTan thanks a lot!! 

You are absolutely right, that global variable is legacy code that is there because in the past  the model was saved outside of the train function (therefore I needed best_acc to be global), but you are right that it should be reset to 0 in the train function. This only affects the case where enc+dec is trained in one pass without stopping the main.py, so I did not debug this case much...

I've updated both lines that you suggest in the code in the repository and now the best model should be saved alright in both enc and dec modes. Thanks again!
